---
title: "Counterfactual Confounder Adjustment for Single-cell Differential Expression Analysis"
author: "Yongjin Park"
---

```{r include=FALSE}
library(tidyverse)
library(data.table)
library(patchwork)
library(R.utils)
require(org.Hs.eg.db)
require(goseq)
source("R/Util-rmd.R")

# Suppress summarise info
options(dplyr.summarise.inform = FALSE)

## document options
fig.dir = "Fig/AD/GO"
dir.create(fig.dir, recursive = TRUE, showWarnings = FALSE)
knitr::opts_chunk$set(fig.path = fig.dir, fig.width = 6, fig.height = 4)
knitr::opts_chunk$set(dev.args = list(bg = "transparent"), echo = FALSE, include = FALSE)
options(stringsAsFactors = FALSE)

## helper functions
.fread <- function(...) fread(..., header=FALSE)
.fwrite <- function(...) fwrite(..., sep="\t", col.names=FALSE, row.names=FALSE)
.read.mat <- function(...) .fread(...) %>% as.matrix
.mkdir <- function(...) dir.create(..., recursive=TRUE, showWarnings=FALSE)
```

UPDATE (Fri 16 Apr 2021 06:58:42 PDT): Use `goseq` instead of a customized script.

```{r}
coding.genes <- fread("data/coding_genes.txt.gz", sep="\t")

run.goseq <- function(.dt, .pheno, .ct, fdr.cutoff, strict.bg = FALSE) {

    if(strict.bg){
        .temp <- .dt[pheno == .pheno & celltype == .ct]
    } else {
        .temp <- .dt[pheno == .pheno]
    }

    .temp <- .temp %>%
        left_join(coding.genes) %>%
        filter(!is.na(ensembl_gene_id))

    .genes <- .temp[, .(ensembl_gene_id)] %>%
        unique %>%
        unlist

    .de.genes <- .temp[celltype == .ct &
                       fdr.cocoa < fdr.cutoff,
                       .(ensembl_gene_id)] %>%
        unique %>%
        unlist

    gene.vector <- as.integer(.genes %in% .de.genes)
    names(gene.vector) <- .genes
    pwf <- goseq::nullp(gene.vector, "hg19", "ensGene", plot.fit=FALSE)
    go.dt <- goseq::goseq(pwf, "hg19", "ensGene", use_genes_without_cat = TRUE) %>%
        as.data.table
}

plot.go.heatmap <- function(go.ct.dt, .celltype.order, fdr.cutoff = .1, .ntop = NULL) {

    go.ct.dt <- go.ct.dt %>%
        mutate(pv = over_represented_pvalue) %>% 
        mutate(qv = p.adjust(pv, "fdr"))

    if(is.null(.ntop)) {
        go.show <- go.ct.dt[qv < fdr.cutoff & numDEInCat > 2,
                            .(term)] %>% unique
    } else {
        go.show <- go.ct.dt[qv < fdr.cutoff & numDEInCat > 2] %>%
            group_by(celltype) %>%
            top_n(.ntop, numDEInCat/numInCat) %>%
            ungroup %>%
            dplyr::select(term) %>%
            unique
    }

    .dt <- merge(go.show, go.ct.dt) %>%
        mutate(row = celltype, col = term, weight = -log10(pv)) %>%
        col.order(.ro = .celltype.order, ret.tab = TRUE) %>%
        as.data.table

    .aes <- aes(y = col, x = row,
                size = -log10(pv),
                fill = log10(numDEInCat))

    .scale.fill <-
        scale_fill_distiller("#genes",
                             palette = "RdPu",
                             direction = 1,
                             labels=function(x) round(10^(x)))

    .scale.size <-
        scale_size_continuous("p-value",
                              breaks = seq(0,10,2),
                              range=c(0, 3),
                              labels=function(x) num.sci(10^(-x)))

    plt <- .gg.plot(.dt, .aes) +
        theme(axis.text.x = element_text(angle=90, vjust=1, hjust=1)) +
        theme(legend.key.width = unit(.2, "lines")) +
        theme(legend.key.height = unit(.5, "lines")) +
        geom_point(pch = 22, stroke=.1) +
        .scale.fill +
        xlab("cell type") + ylab("gene ontology") +
        .scale.size

    .nrow <- length(unique(.dt$term))
    .ncol <- length(unique(.dt$celltype))
    .lab <- max(sapply(.dt$term, nchar))

    list(plt = plt,
         w = (.lab * .055 + .ncol * .05 + .5),
         h = ceiling(.1 * .nrow + .25))
}

save.heatmap <- function(.obj, .file) {
    .mkdir(dirname(.file))
    .gg.save(filename=.file, .obj$plt, width = .obj$w, height = .obj$h)
    return(.obj$plt)
}
```

### Gene ontology enrichment


```{r read_glob_stat, echo=FALSE}
.stat.files <- list.files("result/glob_stat/",
                          pattern=".stat.gz",
                          full.names=TRUE)

stat.dt <-
    lapply(.stat.files, fread, header=TRUE) %>%
    do.call(what=rbind) %>%
    filter(pheno != "age.death") %>%
    mutate(celltype = str_remove(celltype, "subtype-")) %>%
    as.data.table

stat.dt[,
        fdr.cocoa := p.adjust(pv.cocoa, "fdr"),
        by = .(pheno, celltype)]
```


```{r run_goseq}
.mkdir("Tab/AD")
.pheno.ct <-
    stat.dt[, .(pheno, celltype)] %>%
    unique %>%
    as.data.frame

fdr.cutoff <- .2
ntop <- 10

.file <- "Tab/AD/GO_major_types_result.csv.gz"

if(!file.exists(.file)) {
    go.dt <- data.table()

    for(ii in 1:nrow(.pheno.ct)){
        .pheno <- .pheno.ct[ii, "pheno"]
        .ct <- .pheno.ct[ii, "celltype"]

        .go <- run.goseq(stat.dt, .pheno, .ct, fdr.cutoff) %>%
            mutate(pheno = .pheno, celltype = .ct)

        go.dt <- rbind(go.dt, .go)
    }
    fwrite(go.dt, .file)
}
go.dt <- fread(.file)
```

```{r results="asis", include=TRUE}
cat("* [Download](" %&% .file %&% ")")
```

```{r}
.temp <-
    stat.dt %>%
    left_join(coding.genes) %>% 
    filter(!is.na(ensembl_gene_id)) %>% 
    filter(fdr.cocoa < fdr.cutoff) %>%
    as.data.table

count.df <-
    .temp[, .(n = .N), by = .(pheno, celltype)] %>%
    group_by(pheno) %>%
    mutate(weight = log10(n)) %>%
    ungroup()
```

### Using average `r num.int(round(mean(count.df$n)))` $\pm$ `r  num.int(round(sd(count.df$n)))` protein coding genes as a discovery set (FDR 20%)

```{r Fig_gene_count_GO, fig.width=4, fig.height=1.8}
plt <-
    .gg.plot(count.df, aes(x=celltype, y=pheno, fill=weight, label=n)) +
    geom_tile() +
    geom_text(size=3) +
    theme(axis.text.x = element_text(angle=90, vjust=0, hjust=0)) +
    scale_x_discrete(position = "top") +
    scale_fill_distiller(palette=4, labels=function(x) round(10^x)) +
    ylab("phenotype") + xlab("cell type")

print(plt)
.file <- fig.dir %&% "/Fig_gene_count.pdf"
.gg.save(.file, plot=plt, width=4, height=1.8)
```

```{r results="asis", include=TRUE}
cat("* [PDF](" %&% .file %&% ")")
```

```{r}
FDR.CUTOFF = .05
NTOP = 10
```

####  FDR cutoff `r num.sci(FDR.CUTOFF)` showing at most top `r NTOP`

```{r results="asis", include=TRUE}
.ct.names <- c("Ex-L2or3", "Ex-L4", "Ex-L5or6", "Ex-L5or6-CC",
               "In-PV", "In-SST", "In-SV2C", "In-VIP",
               "Microglia", "OPC", "Oligo", "Astro")

for(.pheno in unique(go.dt$pheno)) {

    .file <- fig.dir %&% "/Fig_GO_major_" %&% .pheno %&% ".pdf"
    plt <-
        go.dt[pheno == .pheno & numInCat <= 200 & numInCat >= 10] %>%
        plot.go.heatmap(.ct.names, fdr.cutoff = FDR.CUTOFF, .ntop=NTOP) %>%
        save.heatmap(.file)

    cat("* [" %&% .pheno %&% "](" %&% .file %&% ")\n\n")
}
```

